server/src/main/java/org/candlepin/policy/js/autobind/AutobindRules.java

```diff
@@ -68,29 +68,38 @@ public AutobindRules(JsRunner jsRules, ProductCurator productCurator, RulesObjec
         List<Pool> pools, ComplianceStatus compliance, String serviceLevelOverride,
         Set<String> exemptLevels, boolean considerDerived) {
 
+        List<PoolQuantity> bestPools = new ArrayList<PoolQuantity>();
         int poolsBeforeContentFilter = pools.size();
         pools = filterPoolsForV1Certificates(consumer, pools);
         log.debug("pools.size() before V1 certificate filter: {}, after: {}",
             poolsBeforeContentFilter, pools.size());
 
+        List<String> productIdList;
+        if (productIds == null) {
+            productIdList = new ArrayList<String>();
+        }
+        else {
+            productIdList = Arrays.asList(productIds);
+        }
+
         if (pools.size() == 0) {
             if (compliance.getReasons().size() == 0) {
                 log.info("Consumer is compliant and does not require more entitlements.");
             }
             else {
                 Set<String> fullList = new HashSet<String>();
```

Comment:
Since we're converting the array to a list straight away, it looks like we could sidestep some extra copying here (and below) by just using the productIdList directly.

Also, depending on how the installed products are fetched (lazily?), it might be beneficial to wrap the output blocks in a check for that logging level so we avoid an extra query in those cases.